external-crates/move/crates/move-compiler/src/typing/translate.rs

```diff
@@ -2104,213 +2071,651 @@ fn add_field_types<T>(
     })
 }
 
-enum ExpDotted_ {
-    Exp(Box<T::Exp>),
-    TmpBorrow(Box<T::Exp>, Box<Type>),
-    Dot(Box<ExpDotted>, Field, Box<Type>),
+//--------------------------------------------------------------------------------------------------
+// Index Type Resolution
+//--------------------------------------------------------------------------------------------------
+
+// Assumes tvars have already been readied
+fn find_index_funs(context: &mut Context, loc: Loc, ty: &Type) -> Option<IndexSyntaxMethods> {
+    use TypeName_ as TN;
+    use Type_ as T;
+    const UNINFERRED_MSG: &str =
+        "Could not infer the type before index access. Try annotating here";
+    let ty_str = core::error_format(ty, &context.subst);
+    let msg = || {
+        format!(
+            "No valid '{}({})' method found for {}",
+            SyntaxAttribute::SYNTAX,
+            SyntaxAttribute::INDEX,
+            ty_str
+        )
+    };
+    match ty {
+        sp!(_, T::UnresolvedError) => None,
+        sp!(tloc, T::Anything) => {
+            context.env.add_diag(diag!(
+                TypeSafety::UninferredType,
+                (loc, msg()),
+                (*tloc, UNINFERRED_MSG),
+            ));
+            None
+        }
+        sp!(tloc, T::Var(i)) if !context.subst.is_num_var(*i) => {
```

Comment:
Why specialize num_var here?

Reply:
We should be able to have special syntax on numbers? I don't have any at the moment but it feels plausible... fine with punting that for now though

Reply:
My thinking was that, yes: we might want to carve out the number case later. For now, I have removed this.